File: antlr4/LexerAdaptor.cs
using System;
using System.Reflection;
namespace GrammarGrammar
{
    using System.IO;
    using Antlr4.Runtime;
    using Antlr4.Runtime.Misc;

#pragma warning disable CA1012 // Abstract types should not have constructors
    public abstract class LexerAdaptor : Lexer
#pragma warning restore CA1012 // But Lexer demands it - old 
    {
        // I copy a reference to the stream, so It can be used as a Char Stream, not as a IISStream
        readonly ICharStream stream;
        // Tokens are read only so I hack my way
        readonly FieldInfo tokenInput = typeof(CommonToken).GetField("_type", BindingFlags.NonPublic | BindingFlags.Instance);
        protected LexerAdaptor(ICharStream input)
             : base(input, Console.Out, Console.Error)
        {
            stream = input;
        }

        protected LexerAdaptor(ICharStream input, TextWriter output, TextWriter errorOutput)
             : base(input, output, errorOutput)
        {
            stream = input;
        }
        /**
         * Track whether we are inside of a rule and whether it is lexical parser. _currentRuleType==TokenConstants.InvalidType
         * means that we are outside of a rule. At the first sign of a rule name reference and _currentRuleType==invalid, we
         * can assume that we are starting a parser rule. Similarly, seeing a token reference when not already in rule means
         * starting a token rule. The terminating ';' of a rule, flips this back to invalid type.
         *
         * This is not perfect logic but works. For example, "grammar T;" means that we start and stop a lexical rule for
         * the "T;". Dangerous but works.
         *
         * The whole point of this state information is to distinguish between [..arg actions..] and [charsets]. Char sets
         * can only occur in lexical rules and arg actions cannot occur.
         */
        private int CurrentRuleType { get; set; } = TokenConstants.InvalidType;

        protected void handleBeginArgument()
        {
            if (InLexerRule)
            {
                PushMode(ANTLRv4Lexer.LexerCharSet);
                More();
            }
            else
            {
                PushMode(ANTLRv4Lexer.Argument);
            }
        }

        protected void handleEndArgument()
        {
            PopMode();
            if (ModeStack.Count > 0)
            {
                CurrentRuleType = (ANTLRv4Lexer.ARGUMENT_CONTENT);
            }
        }

        protected void handleEndAction()
        {
            PopMode();
            if (ModeStack.Count > 0)
            {
                CurrentRuleType = (ANTLRv4Lexer.ACTION_CONTENT);
            }
        }


        public override IToken NextToken()
        {
            var token = base.NextToken();
            if (Type == ANTLRv4Lexer.ID)
            {
                char firstChar = stream.GetText(Interval.Of(TokenStartCharIndex, TokenStartCharIndex))[0];
                if (char.IsUpper(firstChar))
                {
                    Type = ANTLRv4Lexer.TOKEN_REF;
                    tokenInput.SetValue(token, ANTLRv4Lexer.TOKEN_REF);
                }
                if (char.IsLower(firstChar))
                {

                    Type = ANTLRv4Lexer.RULE_REF;
                    tokenInput.SetValue(token, ANTLRv4Lexer.RULE_REF);
                }

                if (CurrentRuleType == TokenConstants.InvalidType)
                { // if outside of rule def
                    CurrentRuleType = Type; // set to inside lexer or parser rule
                }
            }
            else if (Type == ANTLRv4Lexer.SEMI)
            { // exit rule def
                CurrentRuleType = TokenConstants.InvalidType;
            }

            return token;
        }

        private bool InLexerRule => CurrentRuleType == ANTLRv4Lexer.TOKEN_REF;


        private bool InParserRule => CurrentRuleType == ANTLRv4Lexer.RULE_REF;

    }
}

File: antlr4/LexerAdaptor.js
/*

Adapted from LexerAdaptor.py by Sebastián Gurin

You need to modify ANTLRv4Lexer.js, use it like this:

```
function ANTLRv4Lexer(input) {
  antlr4.Lexer.call(this, input) 
  this._interp = new antlr4.atn.LexerATNSimulator(this, atn, decisionsToDFA, new antlr4.PredictionContextCache());
  return this;
}
ANTLRv4Lexer.prototype = new LexerAdaptor()
ANTLRv4Lexer.prototype.constructor = ANTLRv4Lexer;
```

*/

var antlr4 = require('antlr4/index');

module.exports.LexerAdaptor = class LexerAdaptor extends antlr4.Lexer {

  getCurrentRuleType() {
    if (typeof this._currentRuleType === 'undefined') {
      this._currentRuleType = antlr4.Token.INVALID_TYPE;
    }
    return this._currentRuleType;
  }

  setCurrentRuleType(ruleType) {
    this._currentRuleType = ruleType;
  }

  handleBeginArgument() {
    if (this.inLexerRule()) {
      this.pushMode(require('./ANTLRv4Lexer').ANTLRv4Lexer.LexerCharSet);
      this.more();
    } else {
      this.pushMode(require('./ANTLRv4Lexer').ANTLRv4Lexer.Argument);
    }
  }

  handleEndArgument() {
    this.popMode();
    if (this._modeStack.length > 0) {
      this._type = require('./ANTLRv4Lexer').ANTLRv4Lexer.ARGUMENT_CONTENT;
    }
  }

  handleEndAction() {
    this.popMode();
    if (this._modeStack.length > 0) {
      this._type = require('./ANTLRv4Lexer').ANTLRv4Lexer.ACTION_CONTENT;
    }
  }

  emit() {
    if (this._type == require('./ANTLRv4Lexer').ANTLRv4Lexer.ID) {
      let firstChar = this._input.getText(this._tokenStartCharIndex, this._tokenStartCharIndex);
      if (firstChar[0].match(/[A-Z]/)) {
        this._type = require('./ANTLRv4Lexer').ANTLRv4Lexer.TOKEN_REF;
      } else {
        this._type = require('./ANTLRv4Lexer').ANTLRv4Lexer.RULE_REF;
      }
      if (this._currentRuleType == antlr4.Token.INVALID_TYPE) {
        this._currentRuleType = this._type
      }
    } else if (this._type == require('./ANTLRv4Lexer').ANTLRv4Lexer.SEMI) {
      this._currentRuleType = antlr4.Token.INVALID_TYPE;
    }
    return super.emit();
  }

  inLexerRule() {
    return this._currentRuleType == require('./ANTLRv4Lexer').ANTLRv4Lexer.TOKEN_REF;
  }

  inParserRule() { // not used, but added for clarity
    return this._currentRuleType == require('./ANTLRv4Lexer').ANTLRv4Lexer.RULE_REF;
  }
}

